### PR TITLE
Openstack Cinder CSI Plugin Pack

### PR DESCRIPTION
**Pull request by rigrassm**
Openstack Cinder CSI Plugin Pack

This PR adds a pack that will deploy the Openstack Cinder CSI plugin as a system job to all clients matching the metadata platform fingerprint for Openstack.

Should partially address #19

### CODE REVIEW

**Comment by rigrassm** [comment]
I toyed with the idea of starting a subdirectory for CSI Plugin packs but didn't know if that's an organizational paradigm this repo is planning on following.

If it is, I would be happy to restructure the pack into a `csi/openstack-cinder` pack instead.

File: packs/csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    
+    [[ template "constraints" .csi_openstack_cinder.constraints ]]
+
+    [[- template "vault_config" .csi_openstack_cinder -]]
+
+    task "cinder-node" {
+      driver = "docker"
+      template {
+        data        = <<EOT
+[[ $config := .csi_openstack_cinder.cloud_conf_file ]][[ fileContents $config ]]
```
Comment (tgross) [inline packs/csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl:25]: I could use some clarification on the handling of the `cloud_conf_file` value.  I checked out this branch and ran it and got the following error:

```
$ nomad-pack run .
! Failed To Process Pack

        Error:   failed to render csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl: template:
csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl:25:57: executing
"csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl" at <fileContents $config>:
error calling fileContents: failed to read : open : no such file or directory
        Type:    *errors.errorString
        Context:
                 - Registry Name: dev
                 - Pack Name: csi_openstack_cinder
                 - Pack Ref: dev
                 - Deployment Name: csi_openstack_cinder
```

So it looks like this is required (let's document that!)? So then I copied the `cloud.conf.example` file to cloud.conf` and that got the error I described above for argument quoting.

File: packs/csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  type = system
+
+  group "nodes" {
+
+    restart {
+      attempts = [[ .csi_openstack_cinder.job_restart_config.attempts ]]
+      delay    = [[ .csi_openstack_cinder.job_restart_config.delay ]]
+      mode     = [[ .csi_openstack_cinder.job_restart_config.mode ]]
+      interval = [[ .csi_openstack_cinder.job_restart_config.interval ]]
+    }
```
Comment (tgross) [inline packs/csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl:15]: I get the following error when running this pack:

```
$ nomad-pack run -var=cloud_conf_file=$(pwd)/cloud.conf .
! Failed To Parse Job Specification

        Error:   400 Bad Request
        Type:    client.GenericOpenAPIError
        Context:
                 - Template Name: csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl
                 - Registry Name: dev
                 - Pack Name: csi_openstack_cinder
                 - Pack Ref: dev
                 - Deployment Name: csi_openstack_cinder
```

I then used `nomad-pack render -var=cloud_conf_file=$(pwd)/cloud.conf .` and copied the results to a file (see https://github.com/hashicorp/nomad-pack/issues/163 for the bug of not being able to redirect that to a file), and I get the following:

```
$ nomad job validate ./cinder.nomad
Error getting job struct: Error parsing job file from ./cinder.nomad:
cinder.nomad:14,20-21: Missing newline after argument; An argument definition must end with a newline.
```

The jobspec is:

<details><summary>cinder.nomad</summary>

```hcl
job "csi_openstack_cinder" {
  region = "global"

  datacenters = [
  "dc1"
]

  type = system

  group "nodes" {

    restart {
      attempts = 5
      delay    = 15s
      mode     = delay
      interval = 5m
    }


  constraint {
    attribute = "${attr.platform.aws.placement.availability-zone}"
    value     = "nova"
  }task "cinder-node" {
      driver = "docker"
      template {
        data        = <<EOT
[Global]
auth-url="https://keystone.cloud.example.net:5000/v3"
application-credential-id="f9f4aaf8-3680-11ec-89d6-0242ac110002"
application-credential-secret="imAvwSfkjuzw2ocaieF9t9+vAGRJ3YpBdnQ/+IUeoa0="
        EOT
        destination = "secrets/cloud.conf"
        change_mode = "restart"
      }
      config {
        image = "docker.io/k8scloudprovider/cinder-csi-plugin:latest"

        mount {
            type     = "bind"
            target   = "/etc/config/cloud.conf"
            source   = "./secrets/cloud.conf"
            readonly = false
            bind_options {
              propagation = "rshared"
            }
          }
        args = [
  "/bin/cinder-csi-plugin",
  "-v=3",
  "--endpoint=unix:///csi/csi.sock",
  "--cloud-config=/etc/config/cloud.conf"
]
        privileged = true
      }

      csi_plugin {
        id        = "csi-cinder"
        type      = "node"
        mount_dir = "/csi"
      }
    }
    task "cinder-controller" {
      driver = "docker"
      template {
        data        = <<EOT
[Global]
auth-url="https://keystone.cloud.example.net:5000/v3"
application-credential-id="f9f4aaf8-3680-11ec-89d6-0242ac110002"
application-credential-secret="imAvwSfkjuzw2ocaieF9t9+vAGRJ3YpBdnQ/+IUeoa0="
        EOT
        destination = "secrets/cloud.conf"
        change_mode = "restart"
      }
      config {
        image = "docker.io/k8scloudprovider/cinder-csi-plugin:latest"
        mount {
            type     = "bind"
            target   = "/etc/config/cloud.conf"
            source   = "./secrets/cloud.conf"
            readonly = false
            bind_options {
              propagation = "rshared"
            }
          }

        args = [
  "/bin/cinder-csi-plugin",
  "-v=3",
  "--endpoint=unix:///csi/csi.sock",
  "--cloud-config=/etc/config/cloud.conf"
]

      }

      csi_plugin {
        id        = "csi-cinder"
        type      = "controller"
        mount_dir = "/csi"
      }
    }
  }
}
```

</details>

It looks like the `delay`, `mode`, and `interval` are all strings and need to be quoted:

```suggestion
    restart {
      attempts = [[ .csi_openstack_cinder.job_restart_config.attempts ]]
      delay    = "[[ .csi_openstack_cinder.job_restart_config.delay ]]"
      mode     = "[[ .csi_openstack_cinder.job_restart_config.mode ]]"
      interval = "[[ .csi_openstack_cinder.job_restart_config.interval ]]"
    }
```

File: packs/csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    
+    [[ template "constraints" .csi_openstack_cinder.constraints ]]
+
+    [[- template "vault_config" .csi_openstack_cinder -]]
```
Comment (tgross) [inline packs/csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl:19]:
```suggestion
    [[- template "vault_config" .csi_openstack_cinder ]]
```

The trailing `-` here causes the newline we need before the `task` block to disappear, like so:

```
  constraint {
    attribute = "${attr.platform.aws.placement.availability-zone}"
    value     = "nova"
  }task "cinder-node" {
      driver = "docker"
      template {
        data        = <<EOT
```

This causes this error on the Nomad side:

```
    2021-11-05T18:09:48.261Z [DEBUG] http: request failed: method=POST path=/v1/jobs/parse error="input.hcl:23,4-8: Missing newline after block definition; A block definition must end with a newline." code=400
```

File: packs/csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl [hunk 4]
```diff
@@ -0,0 +1,84 @@
+job [[ template "job_name" . ]] {
+  [[ template "region" . ]]
+
+  datacenters = [[ .csi_openstack_cinder.datacenters | toPrettyJson ]]
+
+  type = system
```
Comment (tgross) [inline packs/csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl:6]:
```suggestion
  type = "system"
```

This needs to be quoted, or the template engine thinks it's an uninterpolated variable and Nomad shows the following error:

```
    2021-11-05T18:12:58.655Z [ERROR] http: request failed: method=POST path="/v1/jobs?namespace=default&region=global"
  error=
  | 1 error occurred:
  |     * Invalid job type: "${system}"
  |
   code=500
```

File: packs/csi_openstack_cinder/README.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+| <a name="input_constraints"></a> [constraints](#input\_constraints) | Constraints to apply to the entire job. | <pre>list(object({<br>    attribute = string<br>    operator  = string<br>    value     = string<br>  }))</pre> | <pre>[<br>  {<br>    "attribute": "${attr.platform.aws.placement.availability-zone}",<br>    "operator": "",<br>    "value": "nova"<br>  }<br>]</pre> | no |
+| <a name="input_job_restart_config"></a> [job\_restart\_config](#input\_job\_restart\_config) | n/a | <pre>object({<br>      attempts = number<br>      delay    = string<br>      mode     = string<br>      interval = string<br>    })</pre> | <pre>{<br>  "attempts": 5,<br>  "delay": "15s",<br>  "interval": "5m",<br>  "mode": "delay"<br>}</pre> | no |
+| <a name="input_cloud_conf_file"></a> [cloud\_conf\_file](#input\_cloud\_conf\_file) | Path to custom cloud.conf file to be mounted to the CSI containers. For reference, see https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/openstack-cloud-controller-manager/using-openstack-cloud-controller-manager.md#global | `string` | `""` | no |
```
Comment (tgross) [inline packs/csi_openstack_cinder/README.md:24]: Looks like this is required (see my comment on the template below)

**Review by tgross — CHANGES_REQUESTED** [review]
@rigrassm this is looking pretty good. Once I fixed the template issues I've commented on here, it runs and registers the CSI plugin, but I don't have an OpenStack VM to test it out on. So I think if we clean that up this will be ok to merge.

(Without that, the tasks just get into a restart loop exiting 0 (!!!!) with the following warning, as expected)

```
W1105 18:33:06.002302       1 main.go:108] Failed to GetOpenStackProvider: Post "https://keystone.cloud.example.net:5000/v3/auth/tokens": dial tcp: lookup keystone.cloud.example.net on 10.0.2.3:53: no such host
```

File: packs/csi_openstack_cinder/variables.hcl [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+variable "csi_plugin_id" {
+  description = "The ID to register the CSI plugin with."
+  type        = string
+  default     = "csi-cinder"
+}
+
+variable "version_tag" {
+  description = "The docker image version. For options, see https://hub.docker.com/r/k8scloudprovider/cinder-csi-plugin"
```
[the remark this answers is not in the document]
Comment (rigrassm) [reply packs/csi_openstack_cinder/variables.hcl:63]: That's odd, if you search hub for the image it takes you to that exact link but when you try to open straight to it, it redirects to the home page for me... Screen Shot 2021-11-09 at 10 15 02 AM

Gonna replace with the link to the github releases page instead.

[the remark this answers is not in the document]
Comment (tgross) [reply packs/csi_openstack_cinder/variables.hcl:63]: Yeah they're doing something weird over there... I just tried it again and it works just fine now. Good call on replacing with the GH release.

File: packs/csi_openstack_cinder/outputs.tpl [hunk 7]
```diff
@@ -0,0 +1,4 @@
+Cinder CSI Plugin successfully deployed.
+
+See the Openstack Cinder CSI documentation for more information:
+https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/cinder-csi-plugin/using-cinder-csi-plugin.md
```
[the remark this answers is not in the document]
Comment (rigrassm) [reply packs/csi_openstack_cinder/outputs.tpl]: All fixed up

File: packs/csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl (hunk 2 above)
Comment (rigrassm) [reply packs/csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl]: Got these fixed up.

File: packs/csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl (hunk 4 above)
Comment (rigrassm) [reply packs/csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl]: Don't know how I overlooked that one lol. All fixed up.

File: packs/csi_openstack_cinder/README.md (hunk 5 above)
Comment (rigrassm) [reply packs/csi_openstack_cinder/README.md]: Yeah, I think I meant to delete the default declaration in this after copy-pasta'ng and just forgot. Removed the default and regenerated the table.

File: packs/csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl (hunk 3 above)
Comment (rigrassm) [reply packs/csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl]: Added

File: packs/csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl (hunk 1 above)
Comment (rigrassm) [reply packs/csi_openstack_cinder/templates/csi_openstack_cinder.nomad.tpl:23]: Made the variable for providing the file path of the cloud config file a required var so this shouldn't be an issue going forward.

File: packs/csi_openstack_cinder/variables.hcl [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+variable "cinder_node_args" {
+  description = "Arguments passed to the Cinder CSI Node docker container"
+  type        = list(string)
+  default = [
+    "/bin/cinder-csi-plugin",
+    "-v=3",
+    "--endpoint=unix:///csi/csi.sock",
+    "--cloud-config=/etc/config/cloud.conf",
+  ]
```
[the remark this answers is not in the document]
Comment (rigrassm) [reply packs/csi_openstack_cinder/variables.hcl]: Good call. In fact, I don't think there are really any reasons to alter the args at all outside of the verbosity level so I just created a variable for that, removed the two args variable declarations, and moved those args to the template.

**Review by tgross — APPROVED** [review]
LGTM! Thanks @rigrassm!

**Comment by mikenomitch** [comment]
Hey @rigrassm, I come bearing good news. We selected this pack as one of the winners of the [Share Your Nomad Pack for Pack Contest](https://www.hashicorp.com/blog/share-your-nomad-packs-for-a-pack-contest). Thanks for the contribution!

You also got the first community contributed pack into the repo with the Promtail pack, so we owe you both a Nomad T-Shirt and a HashiCorp backpack.

If you're interested, please email me at mnomitch (at) hashicorp (dot) com and we can work out the shipping details.

Thanks for the contribution!

**Pull request merged**